ic-agent/src/identity/mod.rs

```diff
@@ -24,19 +26,75 @@ pub struct Signature {
     pub signature: Option<Vec<u8>>,
 }
 
-/// An Identity takes a request id and returns the [Signature]. It knows or
-/// represents the Principal of the sender.
+/// An `Identity` produces [`Signatures`](Signature) for requests or delegations. It knows or
+/// represents the [`Principal`] of the sender.
 ///
-/// Agents are assigned a single Identity object, but there can be multiple
+/// [`Agents`](crate::Agent) are assigned a single `Identity` object, but there can be multiple
 /// identities used.
 pub trait Identity: Send + Sync {
     /// Returns a sender, ie. the Principal ID that is used to sign a request.
     ///
     /// Only one sender can be used per request.
     fn sender(&self) -> Result<Principal, String>;
 
+    /// Produce the public key commonly returned in [`Signature`].
+    ///
+    /// Should only return `None` if `sign` would do the same.
+    fn public_key(&self) -> Option<Vec<u8>>;
+
     /// Sign a request ID derived from a content map.
     ///
     /// Implementors should call `content.to_request_id().signable()` for the actual bytes that need to be signed.
     fn sign(&self, content: &EnvelopeContent) -> Result<Signature, String>;
+
+    /// Sign a delegation to let another key be used to authenticate [`sender`](Identity::sender).
+    ///
+    /// Not all `Identity` implementations support this operation, though all `ic-agent` implementations other than `AnonymousIdentity` do.
+    ///
+    /// Implementors should call `content.signable()` for the actual bytes that need to be signed.
+    fn sign_delegation(&self, content: &Delegation) -> Result<Signature, String> {
+        let _ = content; // silence unused warning
+        Err(String::from("unsupported"))
+    }
+
+    /// Sign arbitrary bytes.
+    ///
+    /// Not all `Identity` implementations support this operation, though all `ic-agent` implementations do.
```

Comment:
Can you explain why some implementation won't have sign_arbitrary? Eventually, we will get the raw bytes from the envelope, right?

Reply:
Please document that. It's not easy for the user to know when to implement `sign_arbitrary` since it's optional.

Reply:
I do not think 'implement all the operations you can' needs explicit documentation.

Reply:
As a consumer of the API, I need some guideline for when I can expect the optional methods to work without checking the individual implementation. Your reply makes sense, and putting it in the doc will help other users too.

Reply:
There is no such general guideline; if an implementation has a reason it cannot support that operation, the reason is specific to the implementation.